### ISSUE TITLE
Bug: Focalboard plugin is generally very slow

### ISSUE THREAD

**Issue by chenilim** [mattermost/focalboard#560]
**Steps to reproduce the behavior:**
1. Install or upgrade to [Mattermost v5.36 RC1](https://community.mattermost.com/core/channels/release-discussion/3w7q4fac7i8fmqdjscf93mn5rr) or later
2. Enable [custom plugins](https://docs.mattermost.com/administration/plugins.html#custom-plugins) for the server
3. Download the focalboard plugin for [0.7.0 RC1](https://github.com/mattermost/focalboard/releases/tag/v0.7.0-rc1)
4. Upload and enable it
5. Open Focalboard, create a board, and a card
6. Create a Date property on the card
7. It takes a few seconds for the UI to update
8. Similar slow behavior for other actions such as:
  * Drag and drop cards
  * Edit card description 

**Expected behavior:**
Performance should be comparable to the Personal Server web app

**See movie:**

https://user-images.githubusercontent.com/46905241/121611610-7fa77a80-ca0d-11eb-854e-6cba3686d3bf.mov

**Platform:**
 - Chrome 91 on macOS 10.15.7
 - Mattermost v5.36 RC1, Focalboard plugin v0.7.0 RC1

**Additional context:**
I've noticed this slow behavior on 2 separate servers. I thought it was my Internet, but using the focalboard-community server works fine without the slow behavior.

**Comment by chenilim** [comment]
@jespino or @mgdelacroix, could you help investigate this? It's a possible showstopper for the 0.6.7 release. Thanks.

fyi @sbishel.

**Comment by chenilim** [comment]
One more observation: Looking at the Chrome network tab, it appears that onChange is repeatedly firing, even when no action is taken, and on a fresh load. This might be causing the slowdown.

**Comment by chenilim** [comment]
It looks like the websocket is repeatedly closing (or failing to connect), which results in OctoListener performing a retry.

**Comment by chenilim** [comment]
Ok, I think I found the problem - NGINX needs to be configured to upgrade the plugin websocket connections, i.e. by adding this section:

```
   location ~ /plugins/focalboard/ws/* {
       proxy_set_header Upgrade $http_upgrade;
       proxy_set_header Connection "upgrade";
       client_max_body_size 50M;
       proxy_set_header Host $http_host;
       proxy_set_header X-Real-IP $remote_addr;
       proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
       proxy_set_header X-Forwarded-Proto $scheme;
       proxy_set_header X-Frame-Options SAMEORIGIN;
       proxy_buffers 256 16k;
       proxy_buffer_size 16k;
       client_body_timeout 60;
       send_timeout 300;
       lingering_timeout 5;
       proxy_connect_timeout 90;
       proxy_send_timeout 300;
       proxy_read_timeout 90s;
       proxy_pass http://backend;
   }
```

This resolved the issue on my server. However, we need to figure out how to resolve this for our customers.
* One way would be to add the above to the Mattermost install docs, and upgrade docs - can you see an alternative to this for the short term?
  * It would be a huge barrier to adoption, at least for upgraders
* Long term, it would be difficult for customers to add a section for every plugin with a websocket. Is it possible for us to piggy back off the main Mattermost websocket? Or move the websocket endpoint to a central location?

**Comment by chenilim** [comment]
Is anyone familiar with Apache, and can confirm the corresponding config for it? Would the following work:
```
 # Set web sockets
  RewriteEngine On
  RewriteCond %{REQUEST_URI} /api/v[0-9]+/(users/)?websocket|/plugins/focalboard/ws/* [NC,OR]
  RewriteCond %{HTTP:UPGRADE} ^WebSocket$ [NC,OR]
  RewriteCond %{HTTP:CONNECTION} ^Upgrade$ [NC]
  RewriteRule .* ws://127.0.0.1:8065%{REQUEST_URI} [P,QSA,L]
  ```
Thanks.

**Comment by chenilim** [comment]
I've cleaned up and published the steps in discussion item #566.

**Comment by chenilim** [comment]
Closing this issue as we now understand the cause, have added documentation, and a mitigation with PR #570.
